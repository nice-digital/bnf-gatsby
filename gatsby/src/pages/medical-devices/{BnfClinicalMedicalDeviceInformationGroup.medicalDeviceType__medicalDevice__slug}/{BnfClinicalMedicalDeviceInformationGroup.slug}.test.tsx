import { useLocation } from "@reach/router";
import { render, waitFor, screen, within } from "@testing-library/react";
import userEvent from "@testing-library/user-event";

import CMPIPage, {
	type CMPIPageProps,
} from "./{BnfClinicalMedicalDeviceInformationGroup.slug}";

const props: CMPIPageProps = {
	data: {
		bnfClinicalMedicalDeviceInformationGroup: {
			allergyAndCrossSensitivity: null,
			complianceStandards: null,
			deviceDescription: null,
			indicationsAndDose: null,
			medicalDeviceType: {
				slug: "artificial-saliva-products",
				title: "Artificial saliva products",
				medicalDevice: {
					slug: "artificial-saliva-products",
					title: "Artificial saliva products",
				},
				clinicalMedicalDeviceInformationGroups: [
					{
						title: "AS Saliva Orthana® lozenges",
						slug: "as-saliva-orthana-lozenges",
					},
					{
						title: "AS Saliva Orthana® spray",
						slug: "as-saliva-orthana-spray",
					},
					{
						title: "BioXtra® gel",
						slug: "bioxtra-gel",
					},
				],
			},
			patientAndCarerAdvice: null,
			preparations: [],
			prescribingAndDispensingInformation: null,
			professionSpecificInformation: null,
			title: "AS Saliva Orthana® lozenges",
			treatmentCessation: null,
		},
	},
};

describe("CMPIPage", () => {
	describe("SEO", () => {
		it("should set page title from medical type title", async () => {
			render(<CMPIPage {...props} />);
			await waitFor(() => {
				expect(document.title).toStartWith(
					"AS Saliva Orthana® lozenges | Artificial saliva products | Medical devices | "
				);
			});
		});

		it("should set meta description", async () => {
			render(<CMPIPage {...props} />);
			await waitFor(() => {
				expect(
					// eslint-disable-next-line testing-library/no-node-access
					document.querySelector("meta[name=description]")
				).toHaveAttribute(
					"content",
					"This medical devices topic describes the options that are currently recommended for AS Saliva Orthana® lozenges."
				);
			});
		});
	});

	describe("Breadcrumbs", () => {
		it.each([
			["NICE", "https://www.nice.org.uk/"],
			["BNF", "/"],
			["Medical devices", "/medical-devices/"],
			[
				"Artificial saliva products",
				"/medical-devices/artificial-saliva-products/",
			],
		])(
			"should render default '(%s)' breadcrumb",
			(breadcrumbText, expectedHref) => {
				render(<CMPIPage {...props} />);

				const breadcrumbNav = screen.getByRole("navigation", {
					name: "Breadcrumbs",
				});
				const breadcrumb = within(breadcrumbNav).getByRole("link", {
					name: breadcrumbText,
				});

				expect(breadcrumb).toHaveAttribute("href", expectedHref);
			}
		);

		it("should render current page breadcrumb without link", () => {
			render(<CMPIPage {...props} />);

			const breadcrumbNav = screen.getByRole("navigation", {
				name: "Breadcrumbs",
			});
			const currentPageCrumb = within(breadcrumbNav).getByText(
				"AS Saliva Orthana® lozenges",
				{
					selector: "span",
				}
			);
			expect(currentPageCrumb).toBeInTheDocument();
		});
	});

	describe("Page header", () => {
		it("should add content start skip link target id to page header", () => {
			render(<CMPIPage {...props} />);
			const heading1 = screen.getByRole("heading", {
				level: 1,
			});
			// eslint-disable-next-line testing-library/no-node-access
			expect(heading1.parentElement).toHaveAttribute("id", "content-start");
		});

		it("should render heading 1 with current page title", () => {
			render(<CMPIPage {...props} />);
			const heading1 = screen.getByRole("heading", {
				level: 1,
			});
			expect(heading1).toHaveTextContent("AS Saliva Orthana® lozenges");
		});
	});

	describe("Stacked nav", () => {
		it("should not render stacked nav when there are no sibling device types", () => {
			render(
				<CMPIPage
					data={{
						bnfClinicalMedicalDeviceInformationGroup: {
							...props.data.bnfClinicalMedicalDeviceInformationGroup,
							medicalDeviceType: {
								...props.data.bnfClinicalMedicalDeviceInformationGroup
									.medicalDeviceType,
								clinicalMedicalDeviceInformationGroups: [],
							},
						},
					}}
				/>
			);

			expect(
				screen.queryByText("AS Saliva Orthana® lozenges", {
					selector: ".stacked-nav a",
				})
			).toBeNull();
		});

		it("should render stacked nav when there are sibling device types", () => {
			render(<CMPIPage {...props} />);
			expect(
				screen.getByText("Artificial saliva products", {
					selector: ".stacked-nav a",
				})
			).toBeInTheDocument();
		});

		it("should render nav link to parent medical device", () => {
			render(<CMPIPage {...props} />);
			const stackedNav = screen.getByRole("navigation", {
				name: "Artificial saliva products",
			});
			expect(
				within(stackedNav).getByRole("link", {
					name: "Artificial saliva products",
				})
			).toHaveAttribute("href", "/medical-devices/artificial-saliva-products/");
		});

		it.each([
			["AS Saliva Orthana® lozenges", "as-saliva-orthana-lozenges"],
			["AS Saliva Orthana® spray", "as-saliva-orthana-spray"],
			["BioXtra® gel", "bioxtra-gel"],
		])("should render nav link to %s", (name, expectedPath) => {
			render(<CMPIPage {...props} />);
			const stackedNav = screen.getByRole("navigation", {
				name: "Artificial saliva products",
			});
			expect(
				within(stackedNav).getByRole("link", {
					name,
				})
			).toHaveAttribute(
				"href",
				`/medical-devices/artificial-saliva-products/${expectedPath}/`
			);
		});

		it("should highlight current page in stacked nav", () => {
			(useLocation as jest.Mock).mockReturnValue(
				new URL(
					"https://bnf-gatsby-tests.nice.org.uk/medical-devices/artificial-saliva-products/as-saliva-orthana-lozenges/"
				)
			);

			render(<CMPIPage {...props} />);

			const stackedNav = screen.getByRole("navigation", {
				name: "Artificial saliva products",
			});

			expect(
				within(stackedNav).getByRole("link", {
					name: "AS Saliva Orthana® lozenges",
				})
			).toHaveAttribute("aria-current", "true");
		});
	});

	describe("Body", () => {
		describe("Simple pots", () => {
			it.each<
				[
					keyof CMPIPageProps["data"]["bnfClinicalMedicalDeviceInformationGroup"]
				]
			>([
				["deviceDescription"],
				["allergyAndCrossSensitivity"],
				["treatmentCessation"],
				["prescribingAndDispensingInformation"],
				["patientAndCarerAdvice"],
				["professionSpecificInformation"],
				["complianceStandards"],
			])(
				"should create navigate to section link and associated section when %s is present in the feed",
				(sectionName) => {
					render(
						<CMPIPage
							data={{
								bnfClinicalMedicalDeviceInformationGroup: {
									...props.data.bnfClinicalMedicalDeviceInformationGroup,
									[sectionName]: {
										potName: "Test pot",
										slug: "test-pot",
										content: {
											contentFor: "AS Saliva Orthana® lozenges",
											content: "<p>Some content</p>",
										},
									},
								},
							}}
						/>
					);

					const sectionNavButton = screen.getByRole("button", {
						name: "Show Navigate to section",
					});
					userEvent.click(sectionNavButton);

					expect(
						screen.getByRole("link", { name: "Test pot" })
					).toHaveAttribute("href", "#test-pot");
					expect(
						screen.getByRole("region", { name: "Test pot" })
					).toBeInTheDocument();
					expect(
						screen.getByRole("heading", { level: 2, name: "Test pot" })
					).toBeInTheDocument();
					expect(screen.getByText("Some content")).toBeInTheDocument();
				}
			);
		});

		describe("Indications and dose", () => {
			const indicationsAndDose: CMPIPageProps["data"]["bnfClinicalMedicalDeviceInformationGroup"]["indicationsAndDose"] =
				{
					potName: "Indications and dose",
					slug: "indications-and-dose",
					content: {
						contentFor: "AS Saliva Orthana® lozenges",
						doseAdjustments: null,
						doseEquivalence: null,
						extremesOfBodyWeight: null,
						indicationAndDoseGroups: [
							{
								routesAndPatientGroups: [
									{
										routeOfAdministration: "By mouth",
										patientGroups: [
											{
												patientGroup: "adult",
												detailedPatientGroup: "Adult",
												doseStatement:
													"1 lozenge as required, allow to dissolve slowly in the mouth.",
											},
										],
									},
								],
								therapeuticIndications: [
									{
										sctIndication: null,
										sctTherapeuticIntent: null,
										indication:
											"Dry mouth as a result of sicca syndrome (ACBS)",
									},
								],
							},
						],
						pharmacokinetics: null,
						potency: null,
					},
				};

			const indicationsAndDoseProps: CMPIPageProps = {
				data: {
					bnfClinicalMedicalDeviceInformationGroup: {
						...props.data.bnfClinicalMedicalDeviceInformationGroup,
						indicationsAndDose,
					},
				},
			};

			it("should render indications and dose section link and associated section", () => {
				render(<CMPIPage {...indicationsAndDoseProps} />);

				const sectionNavButton = screen.getByRole("button", {
					name: "Show Navigate to section",
				});
				userEvent.click(sectionNavButton);

				expect(
					screen.getByRole("link", { name: indicationsAndDose.potName })
				).toHaveAttribute("href", `#${indicationsAndDose.slug}`);
				expect(
					screen.getByRole("region", { name: indicationsAndDose.potName })
				).toBeInTheDocument();
				expect(
					screen.getByRole("heading", {
						level: 2,
						name: indicationsAndDose.potName,
					})
				).toBeInTheDocument();
			});

			it("should match snapshot for indications and dose section", () => {
				render(<CMPIPage {...indicationsAndDoseProps} />);
				expect(
					screen.getByRole("region", { name: indicationsAndDose.potName })
				).toMatchSnapshot();
			});
		});

		describe("Preparations", () => {
			const preparations: CMPIPageProps["data"]["bnfClinicalMedicalDeviceInformationGroup"]["preparations"] =
					[
						{
							name: "AS Saliva Orthana lozenges",
							manufacturer: "CCMed Ltd",
							ampId: "3804011000001101",
							blackTriangle: false,
							sugarFree: false,
							activeIngredients: [],
							controlledDrugSchedule: null,
							packs: [
								{
									amppId: "3804211000001106",
									size: "30",
									unit: "lozenge",
									nhsIndicativePrice: "£3.04",
									legalCategory: "Not Applicable",
									hospitalOnly: false,
									drugTariff: "Part VIIIA Category C",
									drugTariffPrice: "£3.04",
									colour: null,
								},
							],
						},
					],
				prepsProps: CMPIPageProps = {
					data: {
						bnfClinicalMedicalDeviceInformationGroup: {
							...props.data.bnfClinicalMedicalDeviceInformationGroup,
							preparations,
						},
					},
				};

			it("should render section with accessible name for preps list", () => {
				render(<CMPIPage {...prepsProps} />);
				expect(
					screen.getByRole("region", { name: "Medical device types" })
				).toBeInTheDocument();
			});

			it("should render heading 2 for preps list", () => {
				render(<CMPIPage {...prepsProps} />);
				expect(
					screen.getByRole("heading", {
						level: 2,
						name: "Medical device types",
					})
				).toBeInTheDocument();
			});

			it("should match snapshot for preps list", () => {
				render(<CMPIPage {...prepsProps} />);
				expect(
					screen.getByRole("region", { name: "Medical device types" })
				).toMatchSnapshot();
			});
		});
	});
});
