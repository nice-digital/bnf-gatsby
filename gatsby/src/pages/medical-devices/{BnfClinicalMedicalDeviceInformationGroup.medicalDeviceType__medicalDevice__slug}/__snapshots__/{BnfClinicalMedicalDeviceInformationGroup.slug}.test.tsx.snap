// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CMPIPage Body Indications and dose should match snapshot for indications and dose section 1`] = `
<section
  aria-labelledby="indications-and-dose"
>
  <h2
    id="indications-and-dose"
  >
    Indications and dose
  </h2>
  <section
    aria-labelledby="AS-Saliva-Orthana(r)-lozenges-indications-and-dose"
  >
    <h3
      class="visually-hidden"
      id="AS-Saliva-Orthana(r)-lozenges-indications-and-dose"
    >
      For AS Saliva Orthana® lozenges
    </h3>
    <section
      aria-labelledby="AS-Saliva-Orthana(r)-lozenges-indication-1"
      class="indicationWrapper"
    >
      <h4
        class="indicationHeading"
        id="AS-Saliva-Orthana(r)-lozenges-indication-1"
      >
        <span
          class="indicationText"
        >
          Dry mouth as a result of sicca syndrome (ACBS)
        </span>
        <span
          class="visually-hidden"
        >
           for AS Saliva Orthana® lozenges
        </span>
      </h4>
      <section
        class="routeOfAdministration"
      >
        <h5
          class="routeOfAdministrationHeading"
        >
          By mouth
        </h5>
        <dl
          class="patientGroups"
        >
          <div
            class="patientGroupDose adult"
          >
            <dt
              class="patientGroup"
            >
              Adult
            </dt>
            <dd
              class="doseStatement"
            >
              1 lozenge as required, allow to dissolve slowly in the mouth.
            </dd>
          </div>
        </dl>
      </section>
    </section>
  </section>
</section>
`;

exports[`CMPIPage Body Preparations should match snapshot for preps list 1`] = `
<section
  aria-labelledby="medical-device-types"
>
  <h2
    id="medical-device-types"
  >
    Medical device types
  </h2>
  <ol
    class="prepList"
  >
    <li>
      <details
        class="details  "
      >
        <summary
          class="summary"
          data-tracking="Show"
        >
          <span
            class="toggleLabel label"
          >
            <svg
              aria-hidden="true"
              class="icon"
              fill="none"
              focusable="false"
              height="1em"
              viewBox="0 0 512 512"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M478.152 174c0 10.091-3.525 18.661-10.576 25.712l-186 186c-7.243 7.243-15.909 10.864-26 10.864-10.283 0-18.853-3.621-25.712-10.864l-186-186C36.621 192.853 33 184.283 33 174c0-10.091 3.621-18.757 10.864-26L65 126.576C72.424 119.525 81.09 116 91 116c10.091 0 18.661 3.525 25.712 10.576L255.576 265.44 394.44 126.576c7.051-7.051 15.621-10.576 25.712-10.576 9.909 0 18.576 3.525 26 10.576L467.576 148c7.051 7.424 10.576 16.091 10.576 26z"
                fill="currentColor"
              />
            </svg>
            Show
          </span>
           
          <h3
            class="prepHeading"
          >
            <span
              class="headingText"
            >
              AS Saliva Orthana lozenges
               
              <span
                class="manufacturer"
              >
                CCMed Ltd
              </span>
            </span>
          </h3>
        </summary>
        <ol
          class="packList"
        >
          <li
            class="packItem"
          >
            <dl>
              <div
                class="packDefinitionListItem"
              >
                <dt>
                  Size
                </dt>
                <dd>
                  30
                </dd>
              </div>
              <div
                class="packDefinitionListItem"
              >
                <dt>
                  Unit
                </dt>
                <dd>
                  lozenge
                </dd>
              </div>
              <div
                class="packDefinitionListItem"
              >
                <dt>
                  NHS indicative price
                </dt>
                <dd>
                  £3.04
                </dd>
              </div>
              <div
                class="packDefinitionListItem"
              >
                <dt>
                  Drug tariff
                </dt>
                <dd>
                  Part VIIIA Category C
                </dd>
              </div>
              <div
                class="packDefinitionListItem"
              >
                <dt>
                  Drug tariff price
                </dt>
                <dd>
                  £3.04
                </dd>
              </div>
              <div
                class="packDefinitionListItem"
              >
                <dt>
                  Legal category
                </dt>
                <dd>
                  Not Applicable
                </dd>
              </div>
            </dl>
          </li>
        </ol>
      </details>
    </li>
  </ol>
</section>
`;
